### PR TITLE
Add audit log support to StatBot

### PR DESCRIPTION
**Pull request by ammongit**
This PR adds support for the discord audit log, including a crawler to get past events.

Unfortunately audit logs are a fairly messy item API-wise. The simplest consequence is that because changes may refer to deletions, columns cannot always have foreign keys because the item may not exist anymore. In addition, several columns do not align as well data type-wise, such as `voice_region`, which is a string instead of an `Enum(discord.VoiceRegion)`.

The storage for audit logs are divided between two tables. `audit_log` has a list of every entry, and a lot of the basic data associated with it. But the actual information of what was changed is stored in `audit_log_changes`, since it must store all these fields twice: once for the "before" and again for the "after". Of course, it has a foreign key to the `audit_log` table. Now the `audit_log_changes` table has an insane number of nullable columns, each referring to one possible attribute that could've been changed.

### CODE REVIEW

**Comment by raylu** [comment]
thoughts on http://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.JSON ?

File: statbot/audit_log.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            'deny': deny_perms,
+        }
+
+    def diff_values(self, state):
+        if self.entry.category is None:
+            return None
+
+        if state == AuditLogChangeState.BEFORE:
+            diff = self.entry.before
+        else:
```
Comment (raylu) [inline statbot/audit_log.py]: using `else` to handle enums is how you get bugs like this
https://github.com/Rapptz/discord.py/blob/711dfb83abae5486b70489e0151195ac4f316d52/discord/audit_logs.py#L73

Comment (ammongit) [reply statbot/audit_log.py]: Fair point. I don't understand what you mean by that last statement.

Comment (raylu) [reply statbot/audit_log.py]:
```python
class Animal(enum.Enum):
    cow = 0
    dog = 1

Animal(2)
```
`ValueError: 2 is not a valid Animal`

File: statbot/audit_log.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    def diff_values(self, state):
+        if self.entry.category is None:
+            return None
+
+        if state == AuditLogChangeState.BEFORE:
+            diff = self.entry.before
+        elif state == AuditLogChangeState.AFTER:
+            diff = self.entry.after
+        else:
+            raise ValueError(f"Unknown state: {state}")
```
Comment (raylu) [inline statbot/audit_log.py]: there ain't no need for this `else` if you `alcs = AuditLogChangeState(state)`

**Comment by ammongit** [comment]
Fixes #17

**Pull request merged**